int64 index
int32 x
int32 y
uint8 cost
